src/Sources/AbstractSource.php

```diff
@@ -207,22 +207,36 @@ public function updateAuditTable($source, $changes)
         }
 
         $audit = new Audit();
-        $audit->job_id = $this->getJob()->id;
-        $audit->record_id = $source->id;
-        $audit->source_key = static::getKey();
 
         if ($changes !== null) {
             foreach ($changes as $key => $value) {
+                $this->essentialAuditColumns($source, $audit);
                 $audit->column_name = json_encode($key);
                 $audit->old_value = json_encode($source[$key]);
                 $audit->new_value = json_encode($value);
                 $audit->save();
             }
             return false;
         }
+
+        $this->essentialAuditColumns($source, $audit);
```

Comment:
This still won't quite do what you want.

In this case, you will still only have 1 Audit record. It will just be whatever the *last* change happened to be.
You need to move the `new Audit()` into the loop as well.

Reply:
There you go. Beautiful

Reply:
I see what you mean. I feel like there is a better way to handle this but does it look better now?

Reply:
There may be a marginally better way, but I don't immediately see it. You return early. You only update when you need. You share logic. It passes all the smell tests to me.